docs/generate_docs.sh

```diff
@@ -2,21 +2,49 @@
 
 cd $( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
 
+
 OUTPUT_DIR='../dowhy-docs'
 
-mv source/conf.py source/conf.py.orig
+#
+# Cache existing docs
+#
+[ -d ${OUTPUT_DIR} ] && rm -rf ${OUTPUT_DIR}
+git clone --branch gh-pages git@github.com:py-why/dowhy.git ${OUTPUT_DIR}
+rm -rf ${OUTPUT_DIR}/.git
+
+#
+# Build <0.9 Versions using RTD Theme
+# Note: commenting out so that we can speed up the build by using the pre-built docs for these older versions
+#
+
 cp source/conf-rtd.py source/conf.py
 cp source/_templates/versions-rtd.html source/_templates/versions.html
 
 poetry run sphinx-multiversion --dump-metadata source ${OUTPUT_DIR}
-poetry run sphinx-multiversion source ${OUTPUT_DIR}
 
-mv source/conf.py.orig source/conf.py
+# We expect an error with ret-code=2 when SMV cannot find a version to build
+set +e
+poetry run sphinx-multiversion source ${OUTPUT_DIR}
+retVal=$?
+set -e
+
+if [[ $retVal -ne 0 ]] && [[ $retVal -ne 2 ]]; then
+    echo "error generating documentation"
+    exit $retVal
+fi
+
+#
+# Build >= 0.9 and main branch using Pydata THeme
+#
+cp source/conf-pydata.py source/conf.py
 cp source/_templates/versions-pydata.html source/_templates/versions.html
 
 poetry run sphinx-multiversion --dump-metadata source ${OUTPUT_DIR}
 poetry run sphinx-multiversion source ${OUTPUT_DIR}
```

Comment:
I just realized that we will still have this problem that if we add a new notebook on `main` which uses a new feature from `main`, that this will break, because it runs against a release instead of `main`. Can't we instruct poetry to install DoWhy from the current directory when generating the docs?

Reply:
The approach here is that we don't rebuild released versions. We only rebuild main and new releases using the version-filter in the config files.

Reply:
But at one point you need to build a released version. How do you do this then?

And the other issue also still stands, no? I.e., how do you make sure the notebooks run against `main` and not what's installed from pypi. At least, currently this is why the docs build is failing as documented here: https://github.com/py-why/dowhy/issues/742#issuecomment-1302347510.